chromadb/auth/__init__.py

```diff
@@ -1,449 +1,238 @@
-        )
-
-
-class AuthzDynamicParams:
-    @staticmethod
-    def from_function_name(**kwargs: Any) -> Callable[..., str]:
-        return partial(lambda **kwargs: kwargs["function"].__name__, **kwargs)
-
-    @staticmethod
-    def from_function_args(**kwargs: Any) -> Callable[..., str]:
-        return partial(
-            lambda **kwargs: kwargs["function_args"][kwargs["arg_num"]], **kwargs
-        )
-
-    @staticmethod
-    def from_function_kwargs(**kwargs: Any) -> Callable[..., str]:
-        return partial(
-            lambda **kwargs: kwargs["function_kwargs"][kwargs["arg_name"]], **kwargs
-        )
-
-    @staticmethod
-    def dict_from_function_kwargs(**kwargs: Any) -> Callable[..., Dict[str, Any]]:
-        return partial(
-            lambda **kwargs: {
-                k: kwargs["function_kwargs"][k] for k in kwargs["arg_names"]
-            },
-            **kwargs,
-        )
-
-
-@dataclass
-class AuthzAction:
-    id: str
-    attributes: Optional[Dict[str, Any]] = None
-
-
-@dataclass
-class AuthorizationContext:
-    user: AuthzUser
-    resource: AuthzResource
-    action: AuthzAction
+    """
+    The resource being accessed in an authorization request.
+    """
+    tenant: Optional[str]
+    database: Optional[str]
+    collection: Optional[str]
 
 
 class ServerAuthorizationProvider(Component):
-    def __init__(self, system: System) -> None:
-        super().__init__(system)
-
-    @abstractmethod
-    def authorize(self, context: AuthorizationContext) -> bool:
-        pass
-
-
-class AuthorizationRequestContext(EnforceOverrides, ABC, Generic[T]):
-    @abstractmethod
-    def get_request(self) -> T:
-        ...
-
-
-class ChromaAuthzMiddleware(Component, Generic[T, S]):
-    def __init__(self, system: System) -> None:
-        super().__init__(system)
-
-    @abstractmethod
-    def pre_process(self, request: AuthorizationRequestContext[S]) -> None:
-        ...
-
-    @abstractmethod
-    def ignore_operation(self, verb: str, path: str) -> bool:
-        ...
-
-    @abstractmethod
-    def instrument_server(self, app: T) -> None:
-        ...
-
+    """
+    ServerAuthorizationProvider is responsible for authorizing requests. If a
+    ServerAuthorizationProvider is configured, it will be called by the server
+    to authorize requests. If no ServerAuthorizationProvider is configured, all
+    requests will be authorized.
 
-class ServerAuthorizationConfigurationProvider(Component, Generic[T]):
+    ServerAuthorizationProvider should raise an exception if the request is not
+    authorized.
+    """
     def __init__(self, system: System) -> None:
         super().__init__(system)
 
     @abstractmethod
-    def get_configuration(self) -> T:
+    def authorize_or_raise(self,
+                           user: UserIdentity,
+                           action: AuthzAction,
+                           resource: AuthzResource) -> None:
         pass
+
+    def read_config_or_config_file(self) -> List[str]:
```

Comment:
why is the file reading code a part of the base-class - shouldn't there be a basic impl?

Reply:
What do you mean by "basic impl"? This is logic that's used by both classes and conceivably all subclasses to get config from settings, env vars, or a file; we have three options for where it can live:

- In this base class `ServerAuthorizationProvider` which all authz providers will subclass.
- In a subclass of `ServerAuthorizationProvider`, maybe `ServerAuthorizationProvider{Impl, Skeleton, something els}` and then all authz providers subclass _that_ instead.
- In an entirely separate function.

I don't like the second option because having a parent class while only ever intending to have a single subclass seems weird, but between 1 and 3 I have no real preference.

This also applies to `ServerAuthenticationProvider`s.